### ISSUE TITLE
Should 32 byte `Copy` types be pass by reference or value?

### ISSUE THREAD

**Issue by sanket1729** [rust-bitcoin/rust-bitcoin#708]
Creating a new issue to bring in more attention to this.

Pasting some relevant from #707 .

Sanket1729: 
```
I think we should only pass by value when you give up ownership and I see no reason to do it here. And passing by reference when you temporarily relinquish ownership. I know it's tricky when things are Copy, but that's how it's currently implemented for the rest of the address APIs.

As an extereme, I don't think passing an [0u8; 100000] by value makes sense if we just want to access its elements without any ownership.
```

@Kixunil 
```
To the best of my knowledge, Copy types should be passed by value and it should be the job of the compiler to figure out whether using a pointer internally is better. If it doesn't do that, I'd consider it a compiler bug.

I'd agree with making exception for Option<&Large> as I don't expect the compiler to be able to also do null optimization.

If you have some resources that indicate something else I'd be happy to learn.
```

Sanket1729:
```
My benchmarks say that it does not matter. They are very close in performance and within error of measurement to say if it really matters or not. However semantically, it makes sense to not consume the data because none of the underlying methods require references.

I would 100% support this provided we have underlying rust-secp APIs that require passing by value. What do you think?
```

It would be good to get some consensus on this because this is kinda a blocker on the upcoming taproot release.

**Comment by Kixunil** [comment]
I would swear I've seen some guidelines saying not to use references for `Copy` types but can't find it now. FTR, I don't consider taking by value for `Copy` types "consuming"

**Comment by dpc** [comment]
> I think we should only pass by value when you give up ownership and I see no reason to do it here.

`Copy` types are not being "moved" when passed by value. Caller can still use their copy, and copying instead of moving is the whole meaning of that trait.

There is a possibility of compiler compiling pass-by-copy as an actual copy, but in practice compiler will elide such things and 32B value is so tiny that a copy might even be faster for some reasons, and the magic that modern cpus do under the hood makes the whole performance considerations pointless unless there's some specific hot loop with benchmarks suggesting otherwise.

Given all that the semantically simplest approach is best, which would be by-value, because it avoids `&`s and `*`s and pleasing borrowck.

**Comment by sanket1729** [comment]
>  FTR, I don't consider taking by value for Copy types "consuming". 

Sorry about loosely using the term. After today's research and some benchmarks, I am also of a similar opinion. 

cc @apoelstra @TheBlueMatt who I know might have opinions on this.

**Comment by elichai** [comment]
Haven't formed a strong opinion yet, but some useful information:
https://rust-lang.github.io/rust-clippy/master/#large_types_passed_by_value
https://rust-lang.github.io/rust-clippy/master/#trivially_copy_pass_by_ref

**Comment by sanket1729** [comment]
thanks for the links. Looking at the source code of clippy, it looks like 

For `T: Copy`
- It looks like `&T` is warning if `T < 2 * usize_size`
- Passing by value is a warning if `T` is more than 256 bytes.

**Comment by sanket1729** [comment]
Lastly this. https://github.com/rust-lang/rust/issues/52274, `rustc` does not guarantee anything under the hood. I did some assembly testing :) 

```rust
#[derive(Debug, Clone, Copy)]
struct SuperLargeStruct([usize; 1000]);

#[inline(never)]
fn sum_by_value(s: SuperLargeStruct) -> usize {
    s.0.iter().sum()
}

#[inline(never)]
fn sum_by_ref(s: &SuperLargeStruct) -> usize {
    s.0.iter().sum()
}

fn main() {
    let arr = [0usize; 1000];
    let s = SuperLargeStruct(arr);

    for _ in 0..100 {
        sum_by_value(s);
    }

    for _ in 0..100 {
        sum_by_ref(&s);
    }
}
```

outputs the assembly(only passed the relevant part here)
```asm
	movq	%r14, %rdi
	xorl	%esi, %esi
	callq	*memset@GOTPCREL(%rip)
	movl	$100, %ebp
	leaq	8008(%rsp), %rbx
 # This memcpy is called in the loop
	movq	memcpy@GOTPCREL(%rip), %r15
	.p2align	4, 0x90
.LBB7_1:
	movl	$8000, %edx
	movq	%rbx, %rdi
	movq	%r14, %rsi
	callq	*%r15
	movq	%rbx, %rdi
	callq	_ZN13rust_cp_tests12sum_by_value17hdce064de91892a22E
	addl	$-1, %ebp
	jne	.LBB7_1
	movl	$100, %ebp
 # no memcpy here in case of reference
	leaq	8(%rsp), %rbx
	.p2align	4, 0x90
.LBB7_3:
	movq	%rbx, %rdi 
	callq	_ZN13rust_cp_tests10sum_by_ref17ha0c5305bbc0ef86dE
```

I have added a couple of comments, to highlight just the relevant part. Clearly, we cannot trust llvm/rustc to optimize away `Copy`. It does optimize it away if there is no loop, but as soon as the code gets slightly complicated it fails to optimize memcpys.

**Comment by sanket1729** [comment]
It seems although this is possible, it is technically very difficult. Pasting the relevant part from the thread

> The author of a function can always guarantee predictable cheap argument passing (i.e., not dependent on internals of the function or compiler heuristics) by taking a reference, and I'd say this would be recommended practice for most functions that accept very large and read-only Copy values, regardless of the outcome of this issue.

**Comment by sanket1729** [comment]
This is in --release profile too

**Comment by Kixunil** [comment]
Interesting. Given that 32 bytes is still quite small and benchmarks support it being irrelevant I'd say pass by value for 32-byte types and measure for larger values? Maybe we could also decide about 64 byte values for signatures now?

**Comment by sanket1729** [comment]
I am still unsure. I benched with a crypto-heavy API `Address::p2tr`, which naturally is dominated by other operations, it didn't matter because of the noise around it. 
```
test util::address::benches::bench_addr                                 .. bench:      74,397 ns/iter (+/- 3,561)
```

If I bench the above example (`sum_by_value` and `sum_by_ref`) I can clearly see a very small(~1 ns) difference for 64 bytes.  
```
test util::address::benches::bench_test_ref                             ... bench:       1,721 ns/iter (+/- 165)
test util::address::benches::bench_test_value                           ... bench:       2,660 ns/iter (+/- 582)
```
We need to divide this by 1000 to see the actual difference, so it's really like ~0.5 ns for 32 bytes, 1 ns for 64 bytes and goes up to 28 ns for 2048 bytes. Given that our crypto operations are of the order 74 micro seconds, it's less than 0.01% :) 

But we are just summing over the elements of an array, benchmarks show passing by value is ~2 times worse (because of the extra memcpy). For 2048 byte array, it takes 28 ns (pass by ref) vs ~56 (pass by value). 

There is some difference, but it's really small enough that we should not care for both 32/64 bytes.

**Comment by Kixunil** [comment]
> it didn't matter because of the noise around it.

Actually this is an argument for not caring about that 0.5 ns. If anyone is concerned optimizing crypto operations would help more.

**Comment by sanket1729** [comment]
Agreed, I think, for now, we can keep the new APIs for `schnorr::PublicKey` as pass by value. We can discuss changing other existing APIs in future releases. 

Keeping this issue open for further discussion

**Comment by sanket1729** [comment]
Lastly, I feel really awkward doing this because of the following. 

```rust
fn foo(a: &Obj) {
// None of the methods require `*obj`, why create pass it by value instead of reference
}
```
I think such a change should start in `rust-secp` by changing all the methods to have a pass-by-value and then `rust-bitcoin` would be forced to have all methods as pass-by-value.

**Comment by Kixunil** [comment]
Starting in dependencies does make sense but I wouldn't mind doing it in another way if we expect everything to change eventually.

**Comment by sanket1729** [comment]
> if we expect everything to change eventually.

I would not want to do this unless we have a consensus from the `rust-secp` project. And I think gathering consensus might take really long, which is why I propose going the reference route for now. 

To be clear, I am 100% in favor of changing the APIs to have pass by value, but I would like to see be done in `rust-secp` or at least have some verbal confirmation from other people that it would be done there before we go ahead with this. 

If we are going to change even `bitcoin::PublicKey` to be passed by value everywhere, we should just club of the changes whenever we do them into a single major breaking release.

**Comment by Kixunil** [comment]
I'm in favor too but if we're going to do breaking changes may be worth compiling a list of more things related to Rust API guidelines and have 100% idiomatic and if possible stable API. Yeah, huge project.

**Comment by apoelstra** [comment]
To be clear -- the proposal is to change rust-secp to use pass-by-value for pretty much everything? Concept ACK.

I think in the early days I tried using references as a hacky way to reduce the number of copies of secret data that existed in memory. But aside from this being unreliable, I was also inconsistent about it.

**Comment by Kixunil** [comment]
Yes, pass all `Copy` types by value.

**Comment by TheBlueMatt** [comment]
> I think in the early days I tried using references as a hacky way to reduce the number of copies of secret data that existed in memory. But aside from this being unreliable, I was also inconsistent about it.

Seems to me that `SecretKey` shouldn't be `Copy` to begin with. Maybe we just make this consistent instead?

**Comment by Kixunil** [comment]
@TheBlueMatt that doesn't help with security at all. All moves are `memcpy` and there's no guarantee the "originals" will be overwritten. Even more, the optimizer will do all kind of crazy stuff that's pretty much impossible to control.

Thankfully the exact bugs wiping the secret key would try to protect against are already covered with Rusts safety guarantees, so no huge issue here.

**Comment by apoelstra** [comment]
I think I tried making `SecretKey` non-`Copy` and the result was so unergonomic that I gave up.

Also @Kixunil is correct that "all moves are memcpys" although I imagine in some cases the compiler can be smart enough to avoid this if it knows that a copy is semantically a move.

**Comment by dpc** [comment]
If you consider security of secrets, you might want to take a look at https://github.com/stouset/secrets (and possibly other Rust libraries like this) and either consider using it (or something like it) directly, or maybe provide some facilities to make it usable by users of `rust-bitcoin`

**Comment by Kixunil** [comment]
@dpc I'm personally highly skeptical of crates fixing this. Secrets can end up in registers or God-knows-where and this is determined by the compiler optimizations. Unless this is a feature of the compiler itself I think it's more productive to focus on not allowing to read arbitrary memory in the first place: by preventing memory bugs - we're using the right language here. :)

**Comment by dpc** [comment]
My understanding is as follows. There are two separate things to consider: protecting secret from sticking around after use, and protecting secret while not in use.

Crates like this seems best at protecting secrets not in use. `mlock` will avoid secrets being leaked to storage due to swapping, guard pages provide defense against attempts of stealing the secret by some memory unsafety issue. It's not that it would completely prevent anything, but depending on the scenario, and with combinations with other mitigation techniques (e.g. ASLR), could make stealing the secret harder. For this reason I think libraries like this do have (some) value.  Other mechanism in this library seem mostly user-developer helpers, making dumb programming mistakes less likely.

As for protecting secrets while in use, it's more tricky, but IMO also the problem is a bit less severe. While in theory compilers have a freedom to make copies of stuff, in practice they usually will not because copying is additional work, and compilers are going to try to avoid it. Copies in register are not an issue, since due to limited amount of registers, they very quickly get overwritten. Any memory copies would generally be copies to stack, since the compiler will not want to introduce any memory allocations for temporary copy. Stack gets reused a lot, so while it is possible that a non-zeroed copy of a secret would stick around for a while, that shouldn't generally be the case. I wonder if there are any libraries/protection techniques that would use assembly code to explicitly overwrite certain amount of  bytes above the stack pointer. Such a routine could be called after any operations involving secrets to make sure the leftovers on the stack are gone.

**Comment by Kixunil** [comment]
The effort seems higher than the benefit to me because we already use memory safe language. It's marginal value essentially.

**Comment by TheBlueMatt** [comment]
> @TheBlueMatt that doesn't help with security at all. All moves are memcpy and there's no guarantee the "originals" will be overwritten.

Yes, I'm not under any illusion that rust is somehow safe in this regard, however, that doesn't mean that we should Copy secret keys. You can still prefer to take them by reference in our own APIs and, more importantly, make users think about spewing copies of keys everywhere by making them type clone. That said, really, it's not a huge deal, I just figured if we're discussing it that seems like a better API than the alternative.

**Comment by tcharding** [comment]
For anyone interested I am slowly pulling individual PRs out of #2404 to get this done.

**Comment by Kixunil** [comment]
I've just noticed one argument wasn't mentioned: things like `keys.map(PublicKey::serialize)` are much nicer for copy types since then you're not dealing with weird borrowing problems and don't have to write a bunch of funny symbols.

**Comment by apoelstra** [comment]
True. Now (since 1.36) have the `.copied()` and `.cloned()` adaptors for both iterators and `Option`s. Curiously while we have `Option::as_ref` I cannot find any equivalent for iterators. Maybe there is a borrowck reason. (Not sure if `.map(|x| &x)` even works.)

**Comment by Kixunil** [comment]
> Curiously while we have `Option::as_ref` I cannot find any equivalent for iterators. Maybe there is a borrowck reason. (Not sure if `.map(|x| &x)` even works.)

`map` obviously cannot work because `x` gets dropped at the end of the closure so you can't return `x`. The ways in which it could be done are limited and require streaming iterators. Then if your item `T: 'static` you just need streaming and you wrap an iterator in `Peekable` and impl streaming on top of it, if `T` can contain references then you also have to pin the iterator which currently doesn't have first-class API.
